### DISCUSSION TITLE
Inconsistent NEWLINE handling

### DISCUSSION THREAD

**Discussion by cathaysia** [pest-parser/pest#1015] [Q&A] [answered]
I noticed that the handling of NEWLINE when the rule appears in a COMMENT is inconsistent with that elsewhere.
How to allow NEWLINE in rules in comment?

here is a simple grammar:

```pest
WHITESPACE = _{ WHITE_SPACE }
identifier = { ASCII_ALPHA+ }
cid = { "//@" ~ identifier }

COMMENT = { cid }
document = { cid | identifier }
```

here is the test case:

input :

```
//@Ab
c
```

parse cid:

> - cid > identifier: "Ab\nc"

parse comment:

> - COMMENT: "//@Ab"

expected:

> - COMMENT: "//@Ab\nc"

**Comment by tomtau** [comment]
Is this related to https://github.com/pest-parser/pest/issues/966 ?

**Reply by tomtau** [reply]
Or this: https://github.com/pest-parser/pest/issues/396 ?

You need to enable the grammar-extras feature to use that fix

**Reply by cathaysia** [reply]
I carefully read the two issues you gave:

#966 seems want to expand COMMENT and WHITESPACE rule, so the inner rules will be exposed.
#396 seems says a bug about rules consume tail whitespace.

Both of these questions are different from my situation:

pest will allow any number of whitespace characters between tokens by WHITESPACE, which means that if I have a rule:

```pest
WHITESPACE = _{ WHITE_SPACE | NEWLINE }
identifier = { ASCII_ALPHA+ }
```

then it will match:

> abc

> ab
> c

> a b
> c

when identifier concated by `~`, WHITESPACE rule works also:

```pest
two_identifier = { identifier ~ "=" ~ identifier }
```

two_identifier will match:

> abc = abc

> ab
> c= 
> a    b   c

when two_identifier appears in COMMENT, WHITESPACE doesn't work:

```pest
COMMENT = { two_identifier }
```

it only can match:

> abc=abc

it cannot match:

> abc=
> abc

**Reply by cathaysia** [reply]
here is the full code about above:

```pest
WHITESPACE = _{ WHITE_SPACE | NEWLINE }
identifier = { ASCII_ALPHA+ }
two_identifier = { identifier ~ "=" ~ identifier }
COMMENT = { two_identifier }
```

as a interesting result, two_identifier cannot match:

> abc=abc

it need a space between `c` and `=`. it seems WHITESPACE rule not work normally.

**Reply by tomtau** [reply]
One thing to keep in mind here is that each `~` inserts implicit `WHITESPACE | COMMENT` in-between that the parser may consume.
You can try the debugger: https://github.com/pest-parser/awesome-pest/blob/main/debugger.md and add all rules if you'd like to see what is being tried at which position.

**Reply by cathaysia** [reply]
> ~ inserts implicit WHITESPACE | COMMENT*

but if a rule appears in COMMENT, and this rule maybe contains WHITESPACE, how can I do it?

**Reply by tomtau** [reply, marked as answer]
Perhaps define different explicit rules for those instead of using the implicit ones and use atomic rules: http://pest.rs/book/grammars/syntax.html#atomic ?

**Reply by cathaysia** [reply]
Explicit definition these whitespace does work, but there are too many places to define. :(

**Reply by cathaysia** [reply]
Would you consider allowing WHITESPACE appears in COMMENT?

**Reply by tomtau** [reply]
pest3 will have a bit different grammar style, so hopefully this shouldn't be an issue
